src/3d/castletransform_physics.inc

```diff
@@ -1113,30 +1125,50 @@ end;
 
 procedure TSceneManagerWorld.Update(const SecondsPassed: Single; var RemoveMe: TRemoveType);
 var
-  PhysicsTimeStep: TFloatTime;
   KraftGravity: TVector3;
+  PhysicsTicksCount:Integer;
+  OldTimeAccumulator: TFloatTime;
 begin
   if EnablePhysics and (FKraftEngine <> nil) then
   begin
+    FrameProfiler.Start(fmUpdatePhysics);
+
     // update FKraftEngine.Gravity
     // TODO: do we really need to be prepared that it changes each frame?
     KraftGravity := -GravityUp * 9.81;
     FKraftEngine.Gravity.Vector := VectorToKraft(KraftGravity);
 
+    PhysicsTicksCount := 0;
+
     if not WasPhysicsStep then
     begin
       FKraftEngine.StoreWorldTransforms;
       FKraftEngine.InterpolateWorldTransforms(0.0);
       WasPhysicsStep := true;
     end else
     begin
-      PhysicsTimeStep := 1.0 / FKraftEngine.WorldFrequency;
       TimeAccumulator := TimeAccumulator + SecondsPassed;
-      while TimeAccumulator >= PhysicsTimeStep do
+      while TimeAccumulator >= FPhysicsTimeStep do
       begin
-        TimeAccumulator := TimeAccumulator - PhysicsTimeStep;
+        TimeAccumulator := TimeAccumulator - FPhysicsTimeStep;
         FKraftEngine.StoreWorldTransforms;
-        FKraftEngine.Step(PhysicsTimeStep);
+        FKraftEngine.Step(FPhysicsTimeStep);
+
+        Inc(PhysicsTicksCount);
+
+        { This helps to avoid spiral of death when physics step get more time than rendering. }
```

Comment:
Can we avoid slowing down the physics?

When trying to avoid the spiral of death, we could do one last step with larger time than `FPhysicsTimeStep`. I realize that using "variable time" with physics is not nice, but maybe it's better than slowing down the simulation? Because in your code we just subtract some time from `TimeAccumulator` without "covering" this time with physics simulation.

Can you try something like this within a loop, and see does it still avoid the "spiral of death" in your tests:

```
FKraftEngine.StoreWorldTransforms;

Inc(PhysicsTicksCount);

if (TimeAccumulator >= FPhysicsTimeStep) and
   (PhysicsProperties.MaxPhysicsTicksPerUpdate > 0) and
   (PhysicsTicksCount = PhysicsProperties.MaxPhysicsTicksPerUpdate - 1) then
begin
  { The last step may cover a larger time than usual FPhysicsTimeStep.
    This way we avoid the "spiral of death" by covering more time
    with a single step, if necessary. }
    
  StepTime := FPhysicsTimeStep * Floor(TimeAccumulator / FPhysicsTimeStep)

  WritelnLog('Physics ticks in TSceneManagerWorld.Update() reached ' + IntTostr(PhysicsTicksCount) + '.');
end else
  StepTime := FPhysicsTimeStep;
  
FKraftEngine.Step(StepTime);
TimeAccumulator := TimeAccumulator - StepTime;
```

I looked at Bullet  code of https://github.com/bulletphysics/bullet3/blob/master/src/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp `stepSimulation`, and they do it just like in your code (they just "forget" about the excessive time, when calculating `clampedSimulationSteps` they simply reject the additional time). So it's possible that my idea has a fundamental flaw and I don't see it :) Feel free to reject it if it doesn't really work :)

Reply:
Hi, I was thinking about this too. But I came to the conclusion that, unfortunately, such a solution can lead to strange bugs in specific conditions. I will write more in the general comment to make it more visible (I think it may be useful in the future - when someone wants to change this behavior).